sunxi 6.12: Switch to v6.12.16, Add HDMI support

The-going opened the pull request:
# Description

- Switch to v6.12.16
- sun50i-h616 & overlays: sort nodes, add overlays.
- Add drm sun4i-de33 driver.
- arm64: dts: Add axp313a.dtsi
- add sun50i-h616-hdmi-phy support
- tcon (timing controller), enable HDMI for H616/H618 SOC
- Add GPU power domain for h616
- Add sun50i-h618-bananapi-m4-berry support

# How Has This Been Tested?


- [x] Test build kernel
- [x] Test work HDMI output on boards: BPI-M64, BPI-M4-Berry
- [x] Test build minimal image for BPI-M4-Berry

pyavitz commented:
I haven't been able to get the M4 Zero to boot above v2024.04  in Armbian for some reason. I figured it was something in the Kernel patching as I am able to do so in my personal builds.

Does this PR have eMMC boot support?

The-going commented:
> M4 Zero

Is the Ethernet working?

The-going commented:
> As for the u-boot, yes I've done all that. The issue for me, is that above U-Boot v2024.04 it won't load the Armbian kernel

I suspect that this is with the 6.6 kernel.
At the boot stage, as soon as the temperature driver is connected and starts working,
it reads a very large temperature value and shuts down the OS.

pyavitz commented:
Maybe some of these [patches](https://github.com/pyavitz/debian-image-builder/tree/feature/patches/uboot/v2025.01/bananapim4zero) will help?

igorpecovnik commented:
You decide when this is ready. Also we need to move CURRENT to 6.12, legacy 6.6, edge 6.13 by the end of 2/2025. Any deep problems preventing this?

The-going commented:
> You decide when this is ready. Also we need to move CURRENT to 6.12, legacy 6.6, edge 6.13 by the end of 2/2025. Any deep problems preventing this?

We can reschedule it at any time.
I would like to do this soon.
But I think it's better to fix it first:
 - The 6.6 core works well, except for the temperature sensors for h6-h618, and temperature control is not possible for them.

 - In the 6.12 kernel, for some reason, HDMI does not work for arm64.
 - There are several patches in the 6.12 kernel that add drivers from legacy kernels.
It seems to me that there are problems there.

 - In the 6.13 kernel, which all the patches have moved to, I have already started making some changes.
But they need to be tested. Therefore, I am first adding support for a device with an h618 processor.

What should I do first?
Probably something that will turn out faster.
I will inform you additionally when it will be possible to safely combine pull requests.

igorpecovnik commented:
> What should I do first?

Kernel 6.12 should be main priority, 6.13 slow.

The-going commented:
It's been a long time since I've read so much primary documentation.
I saw some inconsistency with the user manual of the h616 processor.
This applies to syscon and ethernet nodes.

@TheSnowfield @chraac  @JohnTheCoolingFan 
Please tell me, does Ethernet work for your devices on this 6.12 kernel?

JohnTheCoolingFan commented:
> Please tell me, does Ethernet work for your devices on this 6.12 kernel?

Tried this branch on bigtreetech cb1, ethernet works as usual.

Jan-cpu82 commented:
>     * In the 6.12 kernel, for some reason, HDMI does not work for arm64.
It looks like after sunxi-6.7 the "drivers-hack-for-h616-hdmi-video-output.patch" and several
drv-media-cedrus-... patches were deleted in commit 5838fac as "unused"
As h616/618 hdmi support is still not in mainline this is stil needed in my opinion.
The easiest solution is probably to take the latest mainlining effort patchset for 6.12:
https://patchwork.kernel.org/project/dri-devel/cover/20240929091107.838023-1-ryan@testtoast.com/#26045134

igorpecovnik commented:
> The easiest solution 

Nothing is easy ;) We are swamped with work and there might be other patches that needs adjustments. Any help is greatly appreciated if we want to squeeze 6.12 to Armbian 25.02 current kernel. No HDMI is deal breaker.

The-going commented:
> The easiest solution is probably to take the latest mainlining effort patchset for 6.12:

26 patches! But this is already an option to use. Thanks for the advice.

chraac commented:
> It's been a long time since I've read so much primary documentation. I saw some inconsistency with the user manual of the h616 processor. This applies to syscon and ethernet nodes.
> 
> @TheSnowfield @chraac @JohnTheCoolingFan Please tell me, does Ethernet work for your devices on this 6.12 kernel? No changes to this pull request.

The eth work as expected on my opi zero2w
![2e18db81f91cac610726c889dd6b9cc0](https://github.com/user-attachments/assets/249e438c-5f2e-4113-9cd8-4b59c786db7b)

The-going commented:
> The eth work as expected on my opi zero2w

And I suspect that HDMI is running on your board?

In the settings for your board:
[config/boards/orangepizero2w.csc](https://github.com/armbian/build/blob/a3dac432fc08d852a45f27aa37a16a5c1858008a/config/boards/orangepizero2w.csc#L6-L8)

[patch/u-boot/v2024.04/board_orangepizero2w](https://github.com/armbian/build/tree/main/patch/u-boot/v2024.04/board_orangepizero2w)

Bigtreetech-cb1 and bananapi-m4-(zero, berry) use their own separate patch folders.
So, gentlemen, we are doing the same job for the same type of processor (h616-h618).

As far as I understand from the documentation, some hardware parts of the chip are disabled by default.
In order to turn them on, we need to write/change one or more bits in the corresponding registers.
In the image from the manufacturer, this happens at the download stage using the firmware.

It will be very good if we help each other.
Let's put our general changes for u-boot in one folder and make the same settings in our
configuration files for the boards in the build system.
Today I will try to do this for u-boot version v2025.01.
Let's see what happens.

chraac commented:
> > The eth work as expected on my opi zero2w
> 
> And I suspect that HDMI is running on your board?
> 
> In the settings for your board: [config/boards/orangepizero2w.csc](https://github.com/armbian/build/blob/a3dac432fc08d852a45f27aa37a16a5c1858008a/config/boards/orangepizero2w.csc#L6-L8)
> 
> [patch/u-boot/v2024.04/board_orangepizero2w](https://github.com/armbian/build/tree/main/patch/u-boot/v2024.04/board_orangepizero2w)
> 
> Bigtreetech-cb1 and bananapi-m4-(zero, berry) use their own separate patch folders. So, gentlemen, we are doing the same job for the same type of processor (h616-h618).
> 
> As far as I understand from the documentation, some hardware parts of the chip are disabled by default. In order to turn them on, we need to write/change one or more bits in the corresponding registers. In the image from the manufacturer, this happens at the download stage using the firmware.
> 
> It will be very good if we help each other. Let's put our general changes for u-boot in one folder and make the same settings in our configuration files for the boards in the build system. Today I will try to do this for u-boot version v2025.01. Let's see what happens.

The HDMI was not tested yet, 
did remember in 6.6, we've some path to enable the hdmi and gpu, not only in dts, but also some changes in driver
for uboot, yeah we can share the same folder since its mosly the same

The-going commented:
> The HDMI was not tested yet,

Without any changes, simply connect the HDMI cable from the monitor to the board and turn it on.

The-going commented:
Added the DE33 driver.
At this stage, the compilation is error-free.
The core runs on the h618 platform.
For proper operation, it seems necessary to add compatibility with "allwinner,sun50i-h616-de33-clk"

TheSnowfield commented:
> It's been a long time since I've read so much primary documentation. I saw some inconsistency with the user manual of the h616 processor. This applies to syscon and ethernet nodes.
> 
> @TheSnowfield @chraac @JohnTheCoolingFan Please tell me, does Ethernet work for your devices on this 6.12 kernel? No changes to this pull request.

Sorry, I can't test my sakurapi-3308b, because it was a broken boot on the latest Armbian. I haven't figured out the problem.

The-going commented:
> Sorry, I can't test my sakurapi-3308b

I'm sorry. I thought your interest was sunxi. But your devices are on rockchip chips.

TheSnowfield commented:
> > Sorry, I can't test my sakurapi-3308b
> 
> I'm sorry. I thought your interest was sunxi. But your devices are on rockchip chips.

Haha, never mind. sunxi chip devices will come later :)

The-going commented:
DMESG:
```bash
bananapim4berry:~$ dmesg | grep -iE 'drm|hdmi|gpu'
[    0.098994] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.099054] /soc/tcon-top@6510000: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    0.103915] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.104280] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.104438] /soc/tcon-top@6510000: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    0.106983] /soc/hdmi@6000000: Fixed dependency cycle(s) with /connector
[    0.107066] /connector: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    1.597542] sun4i-drm display-engine: bound 1100000.mixer (ops 0xffff80008119ac98)
[    1.604970] sun4i-drm display-engine: bound 6510000.tcon-top (ops 0xffff80008119fa38)
[    1.610933] sun4i-drm display-engine: bound 6515000.lcd-controller (ops 0xffff800081197a60)
[    1.613434] sun8i-dw-hdmi 6000000.hdmi: Detected HDMI TX controller v2.12a with HDCP (DWC HDMI 2.0 TX PHY)
[    1.621580] sun8i-dw-hdmi 6000000.hdmi: EVENT=plugin
[    1.622717] sun8i-dw-hdmi 6000000.hdmi: registered DesignWare HDMI I2C bus driver
[    3.483673] sun4i-drm display-engine: bound 6000000.hdmi (ops 0xffff800081199d60)
[    3.491742] [drm] Initialized sun4i-drm 1.0.0 for display-engine on minor 0
[    3.498754] sun8i-dw-hdmi 6000000.hdmi: read_hpd result: 1
[    4.107497] sun4i-drm display-engine: [drm] fb0: sun4i-drmdrmfb frame buffer device
[    8.567030] systemd[1]: Starting modprobe@drm.service - Load Kernel Module drm...
[    9.102893] systemd[1]: modprobe@drm.service: Deactivated successfully.
[    9.115476] systemd[1]: Finished modprobe@drm.service - Load Kernel Module drm.
[   40.468588] sun8i-dw-hdmi 6000000.hdmi: EVENT=plugin
[   41.362152] sun8i-dw-hdmi 6000000.hdmi: EVENT=plugin
```
At this stage, HDMI works on two boards.
For the h618, the monitor resolution does not switch to the actual available one.
But it's already working.

@igorpecovnik I think it can be taken if someone approves it.

The-going commented:
Please excuse me.
I will add the bananapi-m4-berry board first to the 6.6 core.

igorpecovnik reviewed (approved):
merge at will

The-going commented:
> merge at will

Please wait another 1 day.

EvilOlaf commented:
Has this tested against Orangepi zero 2 and 3? Seems like they broke at this point. Freeze at boot.
https://paste.armbian.de/anepeveyih.yaml
Take a random chance and built the same kernel with BTF disabled but no difference.

The-going commented:
> Has this tested against Orangepi zero 2 and 3? Seems like they broke at this point. Freeze at boot.
> https://paste.armbian.de/anepeveyih.yaml
> Take a random chance and built the same kernel with BTF disabled but no difference.

@EvilOlaf I do not know the reason for the hangup for
Model: Radxa ROCK 3A

For
**CPU:   Allwinner H616 (SUN50I)
Model: OrangePi Zero2**
the problem is related to a patch that adds "GPU activation to the power domain".

Apparently, it does this incorrectly for h616, h618 processors.
We'll just turn it off until the reasons are clarified.

EvilOlaf commented:
> Model: Radxa ROCK 3A 

wtf? Must have brain damage or simply shared the wrong log url lol. Apologies. It is obviously all about the Zero 2 and not the Rock3A.

The-going commented:
> It is obviously all about the Zero 2

I just tested on bananapi-m4-berry (h618).
I have enabled the GPU node in the DTS and disabled this patch.
On the headless image, I see:
```
leo@bananapim4berry:~$ dmesg | grep -E 'gpu|drm|hdmi'
[    0.092338] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.092400] /soc/tcon-top@6510000: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    0.097504] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.097879] /soc/hdmi@6000000: Fixed dependency cycle(s) with /soc/tcon-top@6510000
[    0.098044] /soc/tcon-top@6510000: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    0.100645] /soc/hdmi@6000000: Fixed dependency cycle(s) with /connector
[    0.100733] /connector: Fixed dependency cycle(s) with /soc/hdmi@6000000
[    1.608251] sun4i-drm display-engine: bound 1100000.mixer (ops 0xffff8000811bd460)
[    1.613283] sun4i-drm display-engine: bound 6510000.tcon-top (ops 0xffff8000811c2200)
[    1.619276] sun4i-drm display-engine: bound 6515000.lcd-controller (ops 0xffff8000811ba228)
[    1.621820] sun8i-dw-hdmi 6000000.hdmi: Detected HDMI TX controller v2.12a with HDCP (DWC HDMI 2.0 TX PHY)
[    1.634011] sun8i-dw-hdmi 6000000.hdmi: registered DesignWare HDMI I2C bus driver
[    1.641683] sun4i-drm display-engine: bound 6000000.hdmi (ops 0xffff8000811bc528)
[    3.485436] [drm] Initialized sun4i-drm 1.0.0 for display-engine on minor 0
[    3.492480] sun4i-drm display-engine: [drm] Cannot find any crtc or sizes
[    3.500406] sun4i-drm display-engine: [drm] Cannot find any crtc or sizes
[    7.876316] systemd[1]: Starting modprobe@drm.service - Load Kernel Module drm...
[    8.198265] systemd[1]: modprobe@drm.service: Deactivated successfully.
[    8.205805] systemd[1]: Finished modprobe@drm.service - Load Kernel Module drm.
[    9.738422] panfrost 1800000.gpu: clock rate = 432000000
[    9.738465] panfrost 1800000.gpu: bus_clock rate = 200000000
[    9.739354] panfrost 1800000.gpu: mali-g31 id 0x7093 major 0x0 minor 0x0 status 0x0
[    9.739369] panfrost 1800000.gpu: features: 00000000,000027f7, issues: 00000000,00000400
[    9.739380] panfrost 1800000.gpu: Features: L2:0x07100206 Shader:0x00000000 Tiler:0x00000209 Mem:0x1 MMU:0x00002821 AS:0xff JS:0x7
[    9.739392] panfrost 1800000.gpu: shader_present=0x1 l2_present=0x1
[    9.745620] [drm] Initialized panfrost 1.3.0 for 1800000.gpu on minor 1
```

Explanations:
If we do this in u-boot only once
[patch/u-boot/v2025.01/board_bananapim4berry/0006-h616-GPU-enable.-Magic-register.patch#L19](https://github.com/armbian/build/blob/d9842bb5e8b8f85e61cedebda3511c36009eb126/patch/u-boot/v2025.01/board_bananapim4berry/0006-h616-GPU-enable.-Magic-register.patch#L19)
and [this patch](https://github.com/armbian/build/blob/main/patch/kernel/archive/sunxi-6.12/patches.armbian/clk-sunxi-ng-h6-r-add-GPU-power-domain.patch) is not applied, the GPU is working.

Apparently, the logic in the patch should not be so trivial.
Let's just remove it altogether.

pyavitz commented:
This patch fixes the GPU hang ups; https://lore.kernel.org/linux-sunxi/20250312232319.25712-1-simons.philippe@gmail.com/T/#t

The-going commented:
> This patch fixes the GPU hang ups; https://lore.kernel.org/linux-sunxi/20250312232319.25712-1-simons.philippe@gmail.com/T/#t
> 
> Tested on the BPI-M4-Zero

Should we add this patch to the existing one or use it only?
